IdentityCore/src/cache/metadata/accountMetadata/MSIDAccountMetadata.m

```diff
@@ -0,0 +1,251 @@
+    return YES;
+}
+
+- (NSURL *)cachedURL:(NSURL *)requestURL instanceAware:(BOOL)instanceAware
+{
+    if (self.signInState != MSIDAccountMetadataStateSignedOut)
+    {
+        NSString *urlMapKey = [self URLMapKey:instanceAware];
+        NSDictionary *urlMap = _internalMap[urlMapKey];
+        
+        return [NSURL URLWithString:urlMap[requestURL.absoluteString]];
+    }
+    
+    return nil;
+}
+
+#pragma - Mark Signed out
+- (void)updateSignInState:(MSIDAccountMetadataState)state
+{
+    _signInState = state;
+    if (state == MSIDAccountMetadataStateSignedOut)
+    {
+        _internalMap = [NSMutableDictionary new];
+    }
+    
+}
+
+- (instancetype)initWithJSONDictionary:(NSDictionary *)json
+                                 error:(__unused NSError * __autoreleasing *)error
+{
+    if (!(self = [super init]))
+    {
+        return nil;
+    }
+    
+    if (!json)
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelWarning,nil, @"Tried to decode an authority map item from nil json");
+        return nil;
+    }
+    
+    _clientId = [json msidStringObjectForKey:MSID_CLIENT_ID_CACHE_KEY];
+    _homeAccountId = [json msidStringObjectForKey:MSID_HOME_ACCOUNT_ID_CACHE_KEY];
+    _internalMap = [[json msidObjectForKey:MSID_AUTHORITY_MAP_CACHE_KEY ofClass:NSDictionary.class] mutableDeepCopy];
+    _signInState = [self accountMetadataStateEnumFromString:[json msidStringObjectForKey:MSID_SIGN_IN_STATE_CACHE_KEY]];
+    
+    return self;
+}
+
+- (NSDictionary *)jsonDictionary
+{
+    NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];
+    
+    dictionary[MSID_CLIENT_ID_CACHE_KEY] = self.clientId;
+    dictionary[MSID_HOME_ACCOUNT_ID_CACHE_KEY] = self.homeAccountId;
+    dictionary[MSID_AUTHORITY_MAP_CACHE_KEY] = self.internalMap;
+    dictionary[MSID_SIGN_IN_STATE_CACHE_KEY] = [self accountMetadataStateStringFromEnum:self.signInState];
+    
+    return dictionary;
+}
+
+- (NSString *)URLMapKey:(BOOL)instanceAware
+{
+    // The subkey is in the format of @"URLMap-key1=value1&key2=value2...",
+    // where key1, key2... are the request parameters that may affect the url mapping.
+    // Currently the only parameter that affects the mapping is instance aware flag.
+    //
+    // Example of subkeys:
+    // "URLMap-" : with all possible keys being their default value repectively. Default
+    //             value of instance_aware is NO, so "URLMap-" represents "URLMap-instance_aware=NO"
+    // "URLMap-instance_aware=YES" : with instance_aware being YES.
+    //
+    // The benefit of such a design is, if we are introducing new parameters what will affect the
+    // mapping, there will be no breaking change to existing clients who don't use the new parameters.
+    
+    return instanceAware ? @"URLMap-instance_aware=YES" : @"URLMap-";
+}
+
+- (NSString *)accountMetadataStateStringFromEnum:(MSIDAccountMetadataState)state
+{
+    if (state < 0 || state >= self.accountMetadataStateEnumString.count) return nil;
```

Comment:
nit: a simple switch might be easier to read here :)

Reply:
changed :)